modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/api/SessionImpl.java

```diff
@@ -0,0 +1,283 @@
+        throw new UnsupportedOperationException("Not implemented yet.");
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public int[] executeBatch(@Nullable Transaction transaction, String dmlQuery, BatchedArguments batch) {
+        throw new UnsupportedOperationException("Not implemented yet.");
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public int[] executeBatch(@Nullable Transaction transaction, Statement dmlStatement, BatchedArguments batch) {
+        throw new UnsupportedOperationException("Not implemented yet.");
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void executeScript(String query, @Nullable Object... arguments) {
+        throw new UnsupportedOperationException("Not implemented yet.");
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public long defaultTimeout(TimeUnit timeUnit) {
+        return timeUnit.convert(timeout, TimeUnit.NANOSECONDS);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public String defaultSchema() {
+        return schema;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public int defaultPageSize() {
+        return pageSize;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public @Nullable Object property(String name) {
+        return props.get(name);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public SessionBuilder toBuilder() {
+        if (!busyLock.enterBusy()) {
+            throw new IgniteSqlException("Session is closed");
+        }
+
+        try {
+            return new SessionBuilderImpl(qryProc, props)
+                    .defaultPageSize(pageSize)
+                    .defaultTimeout(timeout, TimeUnit.NANOSECONDS)
+                    .defaultSchema(schema);
+        } finally {
+            busyLock.leaveBusy();
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public CompletableFuture<AsyncResultSet> executeAsync(@Nullable Transaction transaction, String query, @Nullable Object... arguments) {
+        if (!busyLock.enterBusy()) {
+            return CompletableFuture.failedFuture(new IgniteSqlException("Session is closed."));
+        }
+
+        try {
+            QueryContext ctx = QueryContext.of(transaction, new QueryTimeout(timeout, TimeUnit.NANOSECONDS));
+
+            final CompletableFuture<AsyncSqlCursor<List<Object>>> f = qryProc.querySingleAsync(ctx, schema, query, arguments);
+
+            futsToClose.add(f);
+
+            return f.thenCompose(
+                    cur -> {
+                        futsToClose.remove(f);
+
+                        if (!busyLock.enterBusy()) {
+                            return cur.closeAsync()
+                                    .thenCompose((v) -> CompletableFuture.failedFuture(new IgniteSqlException("Session is closed")));
+                        }
+
+                        try {
+                            cursToClose.add(cur);
+
+                            return cur.requestNextAsync(pageSize)
```

Comment:
Seems, cursor will not be removed from 'cursToClose' if 'requestNextAsync' call or the future fails.

Reply:
Fixed, please take a look